server/src/main/java/io/unitycatalog/server/persist/CommitRepository.java

```diff
@@ -0,0 +1,166 @@
+package io.unitycatalog.server.persist;
+
+import io.unitycatalog.server.model.Commit;
+import io.unitycatalog.server.model.Metadata;
+import io.unitycatalog.server.persist.dao.ColumnInfoDAO;
+import io.unitycatalog.server.persist.dao.CommitDAO;
+import io.unitycatalog.server.persist.dao.PropertyDAO;
+import io.unitycatalog.server.persist.dao.TableInfoDAO;
+import io.unitycatalog.server.persist.utils.HibernateUtils;
+import io.unitycatalog.server.utils.Constants;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.NativeQuery;
+import org.hibernate.query.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.UUID;
+
+public class CommitRepository {
+  private static final CommitRepository INSTANCE = new CommitRepository();
+  private static final Logger LOGGER = LoggerFactory.getLogger(CommitRepository.class);
+  private static final SessionFactory SESSION_FACTORY = HibernateUtils.getSessionFactory();
+
+  // The maximum number of commits per table.
+  public static final Integer MAX_NUM_COMMITS_PER_TABLE = 50;
+  private static final Integer NUM_COMMITS_PER_BATCH = 100;
+
+  public static CommitRepository getInstance() {
+    return INSTANCE;
+  }
+
+  public void saveCommit(Commit commit) {
+    CommitDAO commitDAO = CommitDAO.from(commit);
+    try (Session session = SESSION_FACTORY.openSession()) {
+      Transaction tx = session.beginTransaction();
+      try {
+        session.persist(commitDAO);
+        tx.commit();
+      } catch (Exception e) {
+        tx.rollback();
+        LOGGER.error("Failed to save commit", e);
+        throw e;
+      }
+    }
+  }
+
+  public void backfillCommits(
+      UUID tableId, Long upTo, CommitDAO firstCommit, Long highestCommitVersion) {
+    assert upTo >= firstCommit.getCommitVersion();
+    assert upTo < highestCommitVersion;
+    long numCommitsToDelete = upTo - firstCommit.getCommitVersion() + 1;
+    if (numCommitsToDelete <= 0) {
+      return;
+    }
+    // Retry backfilling 5 times to prioritize cleaning of the commit table and log bugs where there
+    // are more
+    // commits in the table than MAX_NUM_COMMITS_PER_TABLE
+    for (int i = 0; i < 5 && numCommitsToDelete > 0; i++) {
+      numCommitsToDelete -= deleteCommits(tableId, upTo);
```

Comment:
What happens when a concurrent API call backfills some commits?

Reply:
We would be retrying 5 times trying to delete the already deleted commits. We could add another check to find the oldest commit in the table and see if that's > the commit number we want to backfill. But that might add some more latency, what do you think?